### PR TITLE
Guard Against Undefined scopeChain

### LINKED ISSUE #308: Uncaught TypeError: undefined is not a function
Uncaught TypeError: undefined is not a function

**Atom Version**: 0.158.0
**System**: Mac OS X 10.10.2
**Thrown From**: [autocomplete-plus](https://github.com/atom-community/autocomplete-plus) package, v2.1.3

### Steps To Reproduce

1. ...
2. ...

### Stack Trace

```
At /Users/youngking/.atom/packages/autocomplete-plus/lib/fuzzy-provider.coffee:167

TypeError: undefined is not a function
  at FuzzyProvider.module.exports.FuzzyProvider.settingsForScopeDescriptor (/Users/youngking/.atom/packages/autocomplete-plus/lib/fuzzy-provider.coffee:167:27)
  at FuzzyProvider.module.exports.FuzzyProvider.getCompletionsForCursorScope (/Users/youngking/.atom/packages/autocomplete-plus/lib/fuzzy-provider.coffee:175:20)
  at FuzzyProvider.getCompletionsForCursorScope (/Users/youngking/.atom/packages/autocomplete-plus/lib/fuzzy-provider.coffee:1:1)
  at FuzzyProvider.module.exports.FuzzyProvider.findSuggestionsForWord (/Users/youngking/.atom/packages/autocomplete-plus/lib/fuzzy-provider.coffee:152:34)
  at FuzzyProvider.findSuggestionsForWord (/Users/youngking/.atom/packages/autocomplete-plus/lib/fuzzy-provider.coffee:1:1)
  at FuzzyProvider.module.exports.FuzzyProvider.requestHandler (/Users/youngking/.atom/packages/autocomplete-plus/lib/fuzzy-provider.coffee:63:20)
  at FuzzyProvider.requestHandler (/Users/youngking/.atom/packages/autocomplete-plus/lib/fuzzy-provider.coffee:1:1)
  at /Users/youngking/.atom/packages/autocomplete-plus/lib/autocomplete-manager.coffee:128:63
  at Array.map (native)
  at AutocompleteManager.module.exports.AutocompleteManager.getSuggestionsFromProviders (/Users/youngking/.atom/packages/autocomplete-plus/lib/autocomplete-manager.coffee:128:35)
  at AutocompleteManager.getSuggestionsFromProviders (/Users/youngking/.atom/packages/autocomplete-plus/lib/autocomplete-manager.coffee:1:1)
  at AutocompleteManager.module.exports.AutocompleteManager.findSuggestions (/Users/youngking/.atom/packages/autocomplete-plus/lib/autocomplete-manager.coffee:124:6)
  at /Users/youngking/.atom/packages/autocomplete-plus/lib/autocomplete-manager.coffee:1:1

```

/cc @atom/core

### PR DESCRIPTION
**Pull request by joefitzgerald**
- Fixes #308

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -3,15 +3,14 @@ version: "{build}"
 os: Windows Server 2012 R2
 
 install:
-  - appveyor DownloadFile https://atom.io/download/windows -FileName AtomSetup.exe
```
Comment (benogle) [inline appveyor.yml]: I assume this stuff shouldnt be in here

File: lib/provider-metadata.coffee [hunk 2]
```diff
@@ -22,12 +23,14 @@ class ProviderMetadata
       false
 
   shouldDisableDefaultProvider: (scopeChain) ->
+    return false unless scopeChain?
     if @disableDefaultProviderSelectors?
```
Comment (benogle) [inline lib/provider-metadata.coffee]: Maybe just add `scopeChain?` to the if?

**Comment by joefitzgerald** [comment]
@benogle :shipit:

**Comment by benogle** [comment]
Hang on, taking a look at the issue. Trying to figure out why scopeChain would be null

**Comment by joefitzgerald** [comment]
My guess is the cursor is undefined because one of the user's packages is stealing focus.

**Comment by joefitzgerald** [comment]
I haven't been able to repro it - this PR simply addresses the actual error.

**Comment by joefitzgerald** [comment]
Scratch that: https://github.com/atom-community/autocomplete-plus/blob/a7397ab5743a74cf1ea90fcee99f3c210d47f42c/lib/autocomplete-manager.coffee#L116 guards against an undefined cursor.

**Comment by joefitzgerald** [comment]
https://github.com/atom-community/autocomplete-plus/blob/a7397ab5743a74cf1ea90fcee99f3c210d47f42c/lib/autocomplete-manager.coffee#L119 has no guard against an undefined scopeDescriptor.

**Comment by joefitzgerald** [comment]
https://github.com/atom-community/autocomplete-plus/blob/a7397ab5743a74cf1ea90fcee99f3c210d47f42c/lib/autocomplete-manager.coffee#L121 could have:

```coffeescript
return unless bufferPosition? and scopeDescriptor?
```

**Comment by benogle** [comment]
I dont think the scopeChain in this case is null or undefined. I think it's not a stirng -- something that does not have a `replace` method. 

I think https://github.com/atom-community/autocomplete-plus/blob/a7397ab5743a74cf1ea90fcee99f3c210d47f42c/lib/provider-manager.coffee#L37 is the culprit.

```
scopeChain = scopeDescriptor?.getScopeChain?() or scopeDescriptor
```

So if there is no `getScopeChain` method, it just gives the scopeDescriptor. Pass in a string or object, and that will be the `scopeChain` passed down. I did this to support the specs passing in a string, but maybe that needs to be revisited.

**Comment by joefitzgerald** [comment]
Right but that would only manifest itself when the scopeDescriptor is undefined, which we now guard against.

**Comment by benogle** [comment]
But we're implicitly checking for that [here](https://github.com/atom-community/autocomplete-plus/blob/a7397ab5743a74cf1ea90fcee99f3c210d47f42c/lib/provider-manager.coffee#L38) `scopeChain` (and by proxy `scopeDescriptor`) is not null or undefined becuause it's getting passed that check.

**Pull request closed without merging**